### ISSUE TITLE
Attachments not recognized since 6.7.0 in MS outlook client/hotmail/gmail

### ISSUE THREAD

**Issue by lubosdz** [PHPMailer/PHPMailer#2938]
- [x] Make sure you're using the latest version of PHPMailer
- [x] Check that your problem is not dealt with in [the troubleshooting guide](https://github.com/PHPMailer/PHPMailer/wiki/Troubleshooting)
- [ ] Include sufficient code to reproduce your problem

# Problem description

After upgrading to latest PHPMailer 6.8 emails stopped displaying file attachments. By downgrading traced back that issue started with 6.7.0 release. Last working was 6.6.5. Since 6.7.0 attachments will not be recognized in **MS Outlook 2010** client and neither on **hotmail.com** & **gmail.com** (!).

I suspected that issue could have been related to newly added method `stripTrailingBreaks()` but cannot confirm - attachments still not recognized even if the method does not strip off trailing whitespaces.

Anyone experiencing same issue?

Here is commit [comparison 6.6.5 - 6.7](https://github.com/PHPMailer/PHPMailer/compare/v6.6.5...v6.7) - if you have any tip what's wrong pls drop it here, I will check out.

**Comment by Synchro** [comment]
Can you show the SMTP transcript of sending a message that's failing in this way? Have you tried comparing messages created by the two different versions to see what the difference is?

**Comment by lubosdz** [comment]
Hi, I compared generated body and found the issue is caused by  missing MIME preamble `This is a multi-part message in MIME format.` in [this](https://github.com/PHPMailer/PHPMailer/commit/12e5c0c6f6652fd4c5b4e98a80ef3b0b1eb80eb7#diff-7f9ff1e59fad629c10da93f0d676eadc5ec86634c576f9852a41e8dba6f8afadL2836) commit.

Showing attachments properly 6.6.5:

~~~
Content-Type: multipart/mixed;
 boundary="b1_OHfdma8GfBImaUYCc4p12Giwhoxlc43JKJMSmcca4Zw"
Content-Transfer-Encoding: 8bit
This is a multi-part message in MIME format.          <- note this line
                                                      <- note this line
--b1_OHfdma8GfBImaUYCc4p12Giwhoxlc43JKJMSmcca4Zw
Content-Type: multipart/alternative;
 boundary="b2_OHfdma8GfBImaUYCc4p12Giwhoxlc43JKJMSmcca4Zw"

--b2_OHfdma8GfBImaUYCc4p12Giwhoxlc43JKJMSmcca4Zw
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: 8bit
~~~

Attachment not shown 6.8.0:

~~~
Content-Type: multipart/mixed;
 boundary="b1=_BxnV55fkP55Ibm0jXjJItoipFfdZh1Snii1LO2saG5Y"
Content-Transfer-Encoding: 8bit
--b1=_BxnV55fkP55Ibm0jXjJItoipFfdZh1Snii1LO2saG5Y


Content-Type: multipart/alternative;
 boundary="b2=_BxnV55fkP55Ibm0jXjJItoipFfdZh1Snii1LO2saG5Y"

--b2=_BxnV55fkP55Ibm0jXjJItoipFfdZh1Snii1LO2saG5Y
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: 8bit
~~~

After I manually added preamble attachments showed up properly. 
It's sad that fixing one issue causes another one :-\

**Comment by Synchro** [comment]
That is annoying. The preamble is entirely optional in the RFC - anything before the first MIME boundary should be ignored, and otherwise it breaks PHPMailer's DKIM, which is why it was removed. Are you sure those line breaks are exactly as in your message? There should be a blank line after the CTE header.

**Comment by lubosdz** [comment]
@Synchro 
Yes & sorry. [Here](https://i.ibb.co/P6RBV5M/screen-phpmailer-preamble.png) is screen with binary comparison.

I am using quite outdated client MS outlook 2010, not sure how it follows RFC. That could be excuse to ignore this thing. Unfortunately, I also noticed attachments not showing on **gmail.com and hotmail.com** - which should not be ignored. Perhaps you/somebody could double check & confirm that? Strange if nobody noticed that by now ..

**Comment by Synchro** [comment]
That binary view is concerning - I suspect the reason it's working in the old one is not because of the preamble, but because of the position of the line break. Without a blank line before it, the preamble looks like a malformed header, and will probably be ignored, but the line after it will act as a acceptable separator between headers and body. In the 6.8 example, the break occurs *after* the boundary, meaning that the boundary will also be interpreted as a malformed header, but here it's more significant because it results in a broken MIME structure (i.e. the first boundary is effectively missing).

**Comment by zb3** [comment]
@Synchro , what break? Note the line count on the right.

What about setting `$mimepre = static::$LE;`?

**Comment by zb3** [comment]
I checked this and it turns out the line works for me without using this hack, also looking at SMTP code confirms that it should work properly. Gmail can recognize attachments properly.

@lubosdz, what transport were you using? Were you using SMTP?

**Comment by lubosdz** [comment]
@zb3 Thank you for checking .. I am not sure actually .. they were stored first as local `EML` files. This way I can check quickly proper content by opening in a local email client, e.g. MS Outlook. Once opened in Outlook, I would change the target email to gmail or hotmail address and send it (from outlook). Does that count as SMTP service .. ? Does it depend how Outlook connects to hosting mail server (POP3 / SMTP) .. ? Well, if it works for you and everybody, then it is probably OK, but I am not sure how to fix showing those attachments since my implementation has not changed since last 2 yrs and before PhpMailer 6.6.5 it all worked fine.

**Comment by Synchro** [comment]
If you set `SMTPDebug = 2` and watch it sending the message, or grab a message with `getSentMIMEMessage()`, are the breaks in the right place?

I recommend using SMTP to localhost over the default `mail()` transport because it's both faster and safer, and also avoids some issues that are common with the mail function (strange wrapping, missing/duplicate headers, escaping problems).

The best way of keeping a copy of messages you send is to BCC yourself and have your mail server pipe the message back into your database. While it's fiddly to set up, the big upside of this approach is that you are guaranteed that the message is an exact copy of what was really sent, so things like DKIM signatures and message IDs (that might be generated by the server) will be present and correct, which may not be the case with local simulations like `getSentMIMEMesage()` provides.
